File: ListView/app/src/main/java/com/example/listview/FruitAdapter.java
package com.example.listview;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class FruitAdapter extends ArrayAdapter<Fruit> {

    public int resourceId;

    public FruitAdapter(@NonNull Context context, int resource, @NonNull List<Fruit> objects) {
        super(context, resource, objects);
        resourceId = resource;
    }

    static class ViewHolder {
        ImageView fruitImg;
        TextView fruitName;
    }

    @Override
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        View view;
        ViewHolder viewHolder;
        if (null == convertView) {
            view = LayoutInflater.from(getContext()).inflate(resourceId,parent,false);
            viewHolder = new ViewHolder();
            viewHolder.fruitImg = view.findViewById(R.id.fruitImage);
            viewHolder.fruitName =  view.findViewById(R.id.fruitName);
            view.setTag(viewHolder);
        } else {
            view = convertView;
            viewHolder = (ViewHolder) view.getTag();
        }
        Fruit fruit = getItem(position);
        viewHolder.fruitImg.setImageResource(fruit.getImgId());
        viewHolder.fruitName.setText(fruit.getName());
        return view;
    }
}

File: RecyclerViewTest/app/src/main/java/com/example/recyclerviewtest/FruitAdapter.kt
package com.example.recyclerviewtest

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView

class FruitAdapter(val fruitList:List<Fruit>) : RecyclerView.Adapter<FruitAdapter.ViewHolder>() {

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val fruitImage: ImageView = view.findViewById(R.id.fruitImage)
        val fruitName: TextView = view.findViewById(R.id.fruitName)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        var view =
            LayoutInflater.from(parent.context).inflate(R.layout.fruit_item, parent, false)
        var viewHolder = ViewHolder(view)
        viewHolder.itemView.setOnClickListener {
            var position = viewHolder.adapterPosition
            var fruit = fruitList[position]
            Toast.makeText(parent.context,"you clicked view ${fruit.name}",Toast.LENGTH_LONG).show()
        }
        viewHolder.fruitImage.setOnClickListener {
            var position = viewHolder.adapterPosition
            var fruit = fruitList[position]
            Toast.makeText(parent.context,"you clicked image ${fruit.name}",Toast.LENGTH_SHORT).show()
        }

        return viewHolder
    }

    override fun getItemCount(): Int = fruitList.size

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var fruit = fruitList.get(position)
        holder.fruitImage.setImageResource(fruit.imgId)
        holder.fruitName.setText(fruit.name)
    }
}
